### ISSUE TITLE
Missing type definitions for GraphQL and Fastify instrumentations when used through auto instrumentation

### ISSUE THREAD

**Issue by dotboris** [open-telemetry/opentelemetry-js-contrib#1266]
### What version of OpenTelemetry are you using?

<details>
<summary>Full versions listing</summary>

```sh
$ yarn list --pattern @opentelemetry
yarn list v1.22.19
├─ @opentelemetry/api-metrics@0.33.0
├─ @opentelemetry/api@1.2.0
├─ @opentelemetry/auto-instrumentations-node@0.34.0
├─ @opentelemetry/context-async-hooks@1.7.0
├─ @opentelemetry/core@1.7.0
├─ @opentelemetry/instrumentation-amqplib@0.31.0
├─ @opentelemetry/instrumentation-aws-lambda@0.33.1
├─ @opentelemetry/instrumentation-aws-sdk@0.9.3
├─ @opentelemetry/instrumentation-bunyan@0.30.1
├─ @opentelemetry/instrumentation-cassandra-driver@0.31.0
├─ @opentelemetry/instrumentation-connect@0.30.1
├─ @opentelemetry/instrumentation-dataloader@0.2.1
├─ @opentelemetry/instrumentation-dns@0.30.1
├─ @opentelemetry/instrumentation-express@0.31.3
├─ @opentelemetry/instrumentation-fastify@0.30.1
├─ @opentelemetry/instrumentation-fs@0.5.1
├─ @opentelemetry/instrumentation-generic-pool@0.30.0
├─ @opentelemetry/instrumentation-graphql@0.32.0
├─ @opentelemetry/instrumentation-grpc@0.32.0
│  ├─ @opentelemetry/api-metrics@0.32.0
│  └─ @opentelemetry/semantic-conventions@1.6.0
├─ @opentelemetry/instrumentation-hapi@0.30.1
├─ @opentelemetry/instrumentation-http@0.32.0
│  ├─ @opentelemetry/core@1.6.0
│  └─ @opentelemetry/semantic-conventions@1.6.0
├─ @opentelemetry/instrumentation-ioredis@0.32.2
├─ @opentelemetry/instrumentation-knex@0.30.1
├─ @opentelemetry/instrumentation-koa@0.33.0
├─ @opentelemetry/instrumentation-lru-memoizer@0.31.0
├─ @opentelemetry/instrumentation-memcached@0.30.1
├─ @opentelemetry/instrumentation-mongodb@0.32.2
├─ @opentelemetry/instrumentation-mongoose@0.31.1
├─ @opentelemetry/instrumentation-mysql@0.31.2
├─ @opentelemetry/instrumentation-mysql2@0.32.1
├─ @opentelemetry/instrumentation-nestjs-core@0.31.1
├─ @opentelemetry/instrumentation-net@0.30.2
├─ @opentelemetry/instrumentation-pg@0.32.0
├─ @opentelemetry/instrumentation-pino@0.32.1
├─ @opentelemetry/instrumentation-redis-4@0.33.1
├─ @opentelemetry/instrumentation-redis@0.33.1
├─ @opentelemetry/instrumentation-restify@0.30.1
├─ @opentelemetry/instrumentation-router@0.31.0
├─ @opentelemetry/instrumentation-tedious@0.4.1
├─ @opentelemetry/instrumentation-winston@0.30.1
├─ @opentelemetry/instrumentation@0.32.0
│  └─ @opentelemetry/api-metrics@0.32.0
├─ @opentelemetry/propagation-utils@0.29.0
├─ @opentelemetry/propagator-aws-xray@1.1.1
├─ @opentelemetry/propagator-b3@1.7.0
├─ @opentelemetry/propagator-jaeger@1.7.0
├─ @opentelemetry/resources@1.7.0
├─ @opentelemetry/sdk-metrics@0.33.0
├─ @opentelemetry/sdk-node@0.33.0
│  └─ @opentelemetry/instrumentation@0.33.0
├─ @opentelemetry/sdk-trace-base@1.7.0
├─ @opentelemetry/sdk-trace-node@1.7.0
└─ @opentelemetry/semantic-conventions@1.7.0
```

</details>

### What version of Node are you using?

`v18.11.0`

### What did you do?

I setup a skeleton project using the node auto instrumentation with typescript:

<details>
<summary>index.ts</summary>

```typescript
import opentelemetry from '@opentelemetry/sdk-node'
import { getNodeAutoInstrumentations } from '@opentelemetry/auto-instrumentations-node'
import { diag, DiagConsoleLogger, DiagLogLevel } from '@opentelemetry/api'

// For troubleshooting, set the log level to DiagLogLevel.DEBUG
diag.setLogger(new DiagConsoleLogger(), DiagLogLevel.INFO)

const sdk = new opentelemetry.NodeSDK({
  traceExporter: new opentelemetry.tracing.ConsoleSpanExporter(),
  instrumentations: [getNodeAutoInstrumentations()]
})

sdk.start()
```

</details>

<details>
<summary>package.json</summary>

```json
{
  "name": "fastify-issue",
  "version": "1.0.0",
  "main": "index.js",
  "license": "MIT",
  "private": true,
  "dependencies": {
    "@opentelemetry/api": "^1.2.0",
    "@opentelemetry/auto-instrumentations-node": "^0.34.0",
    "@opentelemetry/sdk-node": "^0.33.0"
  },
  "devDependencies": {
    "typescript": "^4.8.4"
  }
}
```

</details>

<details>
<summary>tsconfig.json</summary>

```json
{
  "compilerOptions": {
    "target": "es2016",
    "module": "commonjs",
    "esModuleInterop": true,
    "forceConsistentCasingInFileNames": true,
    "strict": true,
    "skipLibCheck": true
  }
}
```

</details>

Then I ran typechecking with all libraries:

```sh
$ yarn tsc --skipLibCheck false --incremental false
```

### What did you expect to see?

I expected the compilation / typecheck to succeed

### What did you see instead?

```
node_modules/@opentelemetry/instrumentation-fastify/build/src/instrumentation.d.ts:2:46 - error TS2307: Cannot find module 'fastify/types/hooks' or its corresponding type declarations.

2 import type { HookHandlerDoneFunction } from 'fastify/types/hooks';
                                               ~~~~~~~~~~~~~~~~~~~~~

node_modules/@opentelemetry/instrumentation-fastify/build/src/instrumentation.d.ts:3:35 - error TS2307: Cannot find module 'fastify/types/reply' or its corresponding type declarations.

3 import type { FastifyReply } from 'fastify/types/reply';
                                    ~~~~~~~~~~~~~~~~~~~~~

node_modules/@opentelemetry/instrumentation-fastify/build/src/instrumentation.d.ts:4:37 - error TS2307: Cannot find module 'fastify/types/request' or its corresponding type declarations.

4 import type { FastifyRequest } from 'fastify/types/request';
                                      ~~~~~~~~~~~~~~~~~~~~~~~

node_modules/@opentelemetry/instrumentation-graphql/build/src/instrumentation.d.ts:2:36 - error TS2307: Cannot find module 'graphql' or its corresponding type declarations.

2 import type * as graphqlTypes from 'graphql';
                                     ~~~~~~~~~

node_modules/@opentelemetry/instrumentation-graphql/build/src/types.d.ts:2:36 - error TS2307: Cannot find module 'graphql' or its corresponding type declarations.

2 import type * as graphqlTypes from 'graphql';
                                     ~~~~~~~~~


Found 5 errors in 3 files.

Errors  Files
     3  node_modules/@opentelemetry/instrumentation-fastify/build/src/instrumentation.d.ts:2
     1  node_modules/@opentelemetry/instrumentation-graphql/build/src/instrumentation.d.ts:2
     1  node_modules/@opentelemetry/instrumentation-graphql/build/src/types.d.ts:2
```

### Additional context

I believe that this may be related to https://github.com/open-telemetry/opentelemetry-js-contrib/issues/1251 and https://github.com/open-telemetry/opentelemetry-js-contrib/pull/1261

The `graphql` and `fastify` instrumentations don't seem to follow the new type management guidelines.

**Comment by dotboris** [comment]
I'm going to start working on a fix for fastify first.

Also, I'm starting to think that there should be a test or linter rule in place that can catch these kinds of issues. My feeling is that it's very easy to break the type management guidelines without realizing.

**Comment by Flarna** [comment]
> Also, I'm starting to think that there should be a test or linter rule in place that can catch these kinds of issues.

That would be really great.

**Comment by dotboris** [comment]
I'm now starting to look at the `graphql` instrumentation. I'll start a separate PR for that

**Comment by blumamir** [comment]
amazing @dotboris 
Thanks again.

After that, if you have time to review #1261 I would really appreciate your feedback.
Specifically, I am interested to know if the text is clear enough and if there is something not covered.

**Comment by dotboris** [comment]
> After that, if you have time to review #1261 I would really appreciate your feedback. Specifically, I am interested to know if the text is clear enough and if there is something not covered.

@blumamir I'm clocking out for the day. I'll try to find some time tomorrow to review your PR. Thank you for the prompt reviews and the great feedback.

**Comment by dotboris** [comment]
> Also, I'm starting to think that there should be a test or linter rule in place that can catch these kinds of issues.

Regarding this, I've done some thinking and digging.

One option would be to write an integration test that does something like the following for every package in the repository:
- Create a `package.json` with the package as a `dependency` and `typescript` as a `devDependency`
- Run `npm install`
- Create an `index.ts` file that does something long the lines of `import * as allTheThings from '{the package}'`
- Run `tsc --skipLibCheck false --incremental false`

I believe that this would work but I think that it'll be very slow (because of the `npm install`) and very flaky because of all the external dependencies, lockfiles and what not.

I'm thinking that this is the right job for static analysis. 

I've dug around to see if there are some eslint rules that could help here. I found <https://github.com/import-js/eslint-plugin-import/blob/HEAD/docs/rules/no-extraneous-dependencies.md> which looks promising but I don't know how well it plays with typescript the fact that private method don't get exported. I'll give it a shot and see how well it works.

I also found [`depcheck`](https://www.npmjs.com/package/depcheck) which seems to be purpose built for that. I've started playing with it and it's giving me trouble. The configuration of it is a little confusing and I feel that it might not play too well with mono repos.

**Comment by dotboris** [comment]
Regarding the `eslint` rule `import/no-extraneous-dependencies`: It's already in use. I've noticed that there's a new unreleased option called `includeTypes` which might work well here. I'm currently giving it a shot.

**Comment by dotboris** [comment]
I tested out `import/no-extraneous-dependencies` with `includeTypes: true` and it doesn't seem to do the job sadly.

I tried running it on `opentelemetry-instrumentation-fastify` and I get the following error. Note that I'm on `main` and I have the fix for #1267 merged. So I should not see any errors.

```
/.../opentelemetry-js-contrib/plugins/node/opentelemetry-instrumentation-fastify/src/internal-types.ts
  18:1  error  'fastify' should be listed in the project's dependencies, not devDependencies  import/no-extraneous-dependencies
```

I then tried to temporarily patch out the existing error and to recreate the issue from #1267 by making `FastifyInstrumentation._wrapAddHook` public. I didn't get any errors from `eslint` but I should have seen one.

From the documentation, it seems like this rule is meant to achieve what's needed here but it doesn't quite work. I'm not sure if it's because the unreleased version is not quite up to snuff or if it's because it doesn't quite understand what's part of the public api and what's not. It's also possible that the plugin may not quite understand some of the idiosyncrasies of typescript.

I'm pretty much out of ideas here. If you have ideas I'd love to hear them. I don't think that I can put in place a good automation that will ensure that people don't accidentally use instrumented packages in public apis.
